src/Databases/DatabaseOnDisk.cpp

```diff
@@ -690,6 +690,17 @@ ASTPtr DatabaseOnDisk::getCreateQueryFromMetadata(const String & database_metada
     if (ast)
     {
         auto & ast_create_query = ast->as<ASTCreateQuery &>();
+        if (ast_create_query.storage &&
+            ast_create_query.storage->engine->name == "Kafka")
+        {
+          for (auto &it : ast_create_query.storage->settings->changes)
+          {
+            if (it.name == "kafka_sasl_password")
+            {
+              it.value.assignString("xxxxx",5);
+            }
+          }
+        }
```

Comment:
> > as CLickHouse DBA, may be need support many teams and each team's kafka cluster settings are different.
> 
> Please remember that every Kafka table is quite an expensive object. Having too many of them is bad idea. See also `background_message_broker_schedule_pool_size`
> 
> > So if every time when user create kafka engine table need inform ClickHouse DBA to modify config.xml
> > and restart clickhouse-server.
> 
> How many teams and kafka brokers with different credential do you expect to have?
> 
> Did you considered using named collections? See https://github.com/ClickHouse/ClickHouse/pull/31691/files
> 
> That is more flexible/general way, which does not require adding explicitly every particular librdkafka settings to storage parameters.
> 
> I.e. currently you can already do that:
> 
> ```
> <!-- /etc/clickhouse-server/config.d/kafka_preset1.xml -->
> <clickhouse>
>     <named_collections>
>         <kafka_preset1>
>             <kafka_broker_list>broker1:345,broker2:345</kafka_broker_list>
>             <kafka_topic_list>topic</kafka_topic_list>
>             <kafka_group_name>clickhouse_topic_group</kafka_group_name>
>             <kafka>
>                 <security_protocol>sasl_ssl</security_protocol>
>                 <sasl_mechanism>SCRAM-SHA-512</sasl_mechanism>
>                 <sasl_username>user</sasl_username>
>                 <sasl_password>password</sasl_password>>
>                 <auto_offset_reset>smallest</auto_offset_reset>
>                 <ssl_endpoint_identification_algorithm>https</ssl_endpoint_identification_algorithm>
>                 <ssl_ca_location>probe</ssl_ca_location>
>             </kafka>
>         </kafka_preset1>
>     </named_collections>
> </clickhouse>
> ```
> 
> ```sql
> CREATE TABLE kafka
> (
>     `key` UInt64,
>     `value` UInt64
> )
> ENGINE = Kafka(kafka_preset1, kafka_format = 'CSV')
> ```
> 
> But attempt to do
> 
> ```sql
> CREATE TABLE kafka
> (
>     `key` UInt64,
>     `value` UInt64
> )
> ENGINE = Kafka(kafka_preset1, kafka_format = 'CSV', kafka.security_protocol = 'PLAIN')
> ```
> 
> ends up with:
> 
> ```
> Received exception from server (version 22.8.4):
> Code: 115. DB::Exception: Received from localhost:9000. DB::Exception: Unknown setting kafka.security_protocol. Stack trace:
> 
> 0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) in /usr/bin/clickhouse
> 1. DB::BaseSettingsHelpers::throwSettingNotFound(std::__1::basic_string_view<char, std::__1::char_traits<char> >) in /usr/bin/clickhouse
> 2. DB::BaseSettings<DB::KafkaSettingsTraits>::set(std::__1::basic_string_view<char, std::__1::char_traits<char> >, DB::Field const&) in /usr/bin/clickhouse
> 3. bool DB::getExternalDataSourceConfiguration<DB::KafkaSettingsTraits>(std::__1::vector<std::__1::shared_ptr<DB::IAST>, std::__1::allocator<std::__1::shared_ptr<DB::IAST> > > const&, DB::BaseSettings<DB::KafkaSettingsTraits>&, std::__1::shared_ptr<DB::Context const>) in /usr/bin/clickhouse
> 4. ? in /usr/bin/clickhouse
> 5. DB::StorageFactory::get(DB::ASTCreateQuery const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Context>, std::__1::shared_ptr<DB::Context>, DB::ColumnsDescription const&, DB::ConstraintsDescription const&, bool) const in /usr/bin/clickhouse
> 6. DB::InterpreterCreateQuery::doCreateTable(DB::ASTCreateQuery&, DB::InterpreterCreateQuery::TableProperties const&) in /usr/bin/clickhouse
> 7. DB::InterpreterCreateQuery::createTable(DB::ASTCreateQuery&) in /usr/bin/clickhouse
> 8. DB::InterpreterCreateQuery::execute() in /usr/bin/clickhouse
> 9. ? in /usr/bin/clickhouse
> 10. DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) in /usr/bin/clickhouse
> 11. DB::TCPHandler::runImpl() in /usr/bin/clickhouse
> 12. DB::TCPHandler::run() in /usr/bin/clickhouse
> 13. Poco::Net::TCPServerConnection::start() in /usr/bin/clickhouse
> 14. Poco::Net::TCPServerDispatcher::run() in /usr/bin/clickhouse
> 15. Poco::PooledThread::run() in /usr/bin/clickhouse
> 16. Poco::ThreadImpl::runnableEntry(void*) in /usr/bin/clickhouse
> 17. ? in ?
> 18. __clone in ?
> . (UNKNOWN_SETTING)
> ```
> 
> I think that should not be hard to support.


> > as CLickHouse DBA, may be need support many teams and each team's kafka cluster settings are different.
> 
> Please remember that every Kafka table is quite an expensive object. Having too many of them is bad idea. See also `background_message_broker_schedule_pool_size`
> 
> > So if every time when user create kafka engine table need inform ClickHouse DBA to modify config.xml
> > and restart clickhouse-server.
> 
> How many teams and kafka brokers with different credential do you expect to have?
> 
> Did you considered using named collections? See https://github.com/ClickHouse/ClickHouse/pull/31691/files
> 
> That is more flexible/general way, which does not require adding explicitly every particular librdkafka settings to storage parameters.
> 
> I.e. currently you can already do that:
> 
> ```
> <!-- /etc/clickhouse-server/config.d/kafka_preset1.xml -->
> <clickhouse>
>     <named_collections>
>         <kafka_preset1>
>             <kafka_broker_list>broker1:345,broker2:345</kafka_broker_list>
>             <kafka_topic_list>topic</kafka_topic_list>
>             <kafka_group_name>clickhouse_topic_group</kafka_group_name>
>             <kafka>
>                 <security_protocol>sasl_ssl</security_protocol>
>                 <sasl_mechanism>SCRAM-SHA-512</sasl_mechanism>
>                 <sasl_username>user</sasl_username>
>                 <sasl_password>password</sasl_password>>
>                 <auto_offset_reset>smallest</auto_offset_reset>
>                 <ssl_endpoint_identification_algorithm>https</ssl_endpoint_identification_algorithm>
>                 <ssl_ca_location>probe</ssl_ca_location>
>             </kafka>
>         </kafka_preset1>
>     </named_collections>
> </clickhouse>
> ```
> 
> ```sql
> CREATE TABLE kafka
> (
>     `key` UInt64,
>     `value` UInt64
> )
> ENGINE = Kafka(kafka_preset1, kafka_format = 'CSV')
> ```
> 
> But attempt to do
> 
> ```sql
> CREATE TABLE kafka
> (
>     `key` UInt64,
>     `value` UInt64
> )
> ENGINE = Kafka(kafka_preset1, kafka_format = 'CSV', kafka.security_protocol = 'PLAIN')
> ```
> 
> ends up with:
> 
> ```
> Received exception from server (version 22.8.4):
> Code: 115. DB::Exception: Received from localhost:9000. DB::Exception: Unknown setting kafka.security_protocol. Stack trace:
> 
> 0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) in /usr/bin/clickhouse
> 1. DB::BaseSettingsHelpers::throwSettingNotFound(std::__1::basic_string_view<char, std::__1::char_traits<char> >) in /usr/bin/clickhouse
> 2. DB::BaseSettings<DB::KafkaSettingsTraits>::set(std::__1::basic_string_view<char, std::__1::char_traits<char> >, DB::Field const&) in /usr/bin/clickhouse
> 3. bool DB::getExternalDataSourceConfiguration<DB::KafkaSettingsTraits>(std::__1::vector<std::__1::shared_ptr<DB::IAST>, std::__1::allocator<std::__1::shared_ptr<DB::IAST> > > const&, DB::BaseSettings<DB::KafkaSettingsTraits>&, std::__1::shared_ptr<DB::Context const>) in /usr/bin/clickhouse
> 4. ? in /usr/bin/clickhouse
> 5. DB::StorageFactory::get(DB::ASTCreateQuery const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Context>, std::__1::shared_ptr<DB::Context>, DB::ColumnsDescription const&, DB::ConstraintsDescription const&, bool) const in /usr/bin/clickhouse
> 6. DB::InterpreterCreateQuery::doCreateTable(DB::ASTCreateQuery&, DB::InterpreterCreateQuery::TableProperties const&) in /usr/bin/clickhouse
> 7. DB::InterpreterCreateQuery::createTable(DB::ASTCreateQuery&) in /usr/bin/clickhouse
> 8. DB::InterpreterCreateQuery::execute() in /usr/bin/clickhouse
> 9. ? in /usr/bin/clickhouse
> 10. DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) in /usr/bin/clickhouse
> 11. DB::TCPHandler::runImpl() in /usr/bin/clickhouse
> 12. DB::TCPHandler::run() in /usr/bin/clickhouse
> 13. Poco::Net::TCPServerConnection::start() in /usr/bin/clickhouse
> 14. Poco::Net::TCPServerDispatcher::run() in /usr/bin/clickhouse
> 15. Poco::PooledThread::run() in /usr/bin/clickhouse
> 16. Poco::ThreadImpl::runnableEntry(void*) in /usr/bin/clickhouse
> 17. ? in ?
> 18. __clone in ?
> . (UNKNOWN_SETTING)
> ```
> 
> I think that should not be hard to support.

hello @filimonov  many thx reply and reviews, good suggestions for many comments,
AS ClickHouse DBA, they maybe maintain more than one clickhouse db instance, yes, too many kafka tables is heavy for one clickhouse DB instance. But for considering DBA maybe need support their whole company 's AP business, and each business team which  will have one ClickHouse  DB instance, so maybe need support dozens ClickHouse DB instance. So let user can set sasl_settings is better instead of help user to modify the config.xml

And kafka.security_protocol should type kafka_security_protocol, i will added test cases , resolved you comments  then Commit again

At last, thx for your comments.

Reply:
>  kafka.security_protocol should type kafka_security_protocol

actually... not :) See https://github.com/ClickHouse/ClickHouse/blob/62c911924b0866fc6a69517baf79d2a8fbf998dd/src/Storages/Kafka/StorageKafka.cpp#L508-L518 

So config prefix for named collection is `named_collections.<preset_name>.kafka.<librfkafka_setting>`

Named collections allow to push adjustments on the top level, but it seems they don't allow to do that in one sublevel.

> So let user can set sasl_settings is better instead of help user to modify the config.xml

Please remember that there are different ways of Kafka auth, some of them require password, some certificates, some kerberos etc. 

Also there are security concerns, if the password is passed to the table settings it can be easily accessed by `SHOW CREATE TABLE ... `, so option with the config is MUCH safer.

Reply:
> > kafka.security_protocol should type kafka_security_protocol
> 
> actually... not :) See
> 
> https://github.com/ClickHouse/ClickHouse/blob/62c911924b0866fc6a69517baf79d2a8fbf998dd/src/Storages/Kafka/StorageKafka.cpp#L508-L518
> 
> So in case of named collections clickhouse expects to get librdkafka settings from `named_collections.<collection_name>.kafka.<librfkafka_setting_name>` + optionally `named_collections.<collection_name>.kafka_<topic_name>.<librfkafka_setting_name>`
> 
> Named collections allow to push adjustments on the top level of collection, but it seems they don't allow to do that in one sublevel.
> 
> > So let user can set sasl_settings is better instead of help user to modify the config.xml
> 
> Please remember that there are different ways of Kafka auth, some of them require password, some certificates, some kerberos etc.
> 
> Also there are security concerns, if the password is passed to the table settings it can be easily accessed by `SHOW CREATE TABLE ... `, so option with the config is MUCH safer.

understand you said for the safety
so this patch, when user using show create table, the password part will show xxxxx, through the show create table or query the system.tables will not show the password

![image](https://user-images.githubusercontent.com/26053753/189265349-54bc9c79-8235-4b4a-b1e7-9df176cfe237.png)



and this patch already  support different sasl mechanism and security protocol, as a clickhouse DBA, i also maintain more than 50 ClickHouse DB instance, and each of instance normally have 2 ~ 5 Kafka engine table. the supported sasl security mechanism and protocol already can cover all the requirements
and show create table return xxxx can keep the password safe

Thx for you comments

I will try to provide test cases and solved you other comments in recent two days